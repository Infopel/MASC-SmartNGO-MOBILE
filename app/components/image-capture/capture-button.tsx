import { Button } from "components/button/button";
import { Button as NBbtn } from "native-base";
import React from 'react';
interface CaptureButtonProps {
    onPressAdd: () => void;
    onPressPreview: () => void;
}
export function CaptureButton(props: CaptureButtonProps) {
    return <NBbtn.Group isAttached flex='1'>
        <Button tx="common.camera-add" flex='1' onPress={props.onPressAdd} />
        <Button tx="common.camera-preview" flex='1' variant='subtle'  onPress={props.onPressPreview} />
    </NBbtn.Group>
}