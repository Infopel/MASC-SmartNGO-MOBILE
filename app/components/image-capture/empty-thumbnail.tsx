import { Icon } from "components/icon/icon";
import React from "react";

const sizes = {
    sm: '16',
    lg: '48',
}


export interface EmptyThumbnailProps {
    size: keyof typeof sizes

}
export function EmptyThumbnail({ size = 'sm' }: EmptyThumbnailProps) {
    const _size = sizes[size]

    return (
        <Icon name='image-not-supported' type='material' size={_size} />

    )
}