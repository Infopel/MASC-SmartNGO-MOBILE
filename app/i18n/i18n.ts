import * as Localization from "expo-localization"
//@ts-ignore
import {I18n} from "i18n-js/dist/require"
import en from "./en"
import pt from "./pt"
export const i18n =  new I18n({en,pt})
i18n.enableFallback = true
i18n.defaultLocale = "en"
// i18n.missingBehavior = "error";

i18n.locale = Localization.locale || "en"

/**
 * Builds up valid keypaths for translations.
 * Update to your default locale of choice if not English.
 */
type DefaultLocale = typeof en
export type TxKeyPath = RecursiveKeyOf<DefaultLocale>

type RecursiveKeyOf<TObj extends Record<string, any>> = {
  [TKey in keyof TObj & string]: TObj[TKey] extends Record<string, any>
  //@ts-ignore  
  ? `${TKey}` | `${TKey}.${RecursiveKeyOf<TObj[TKey]>}`
    : `${TKey}`
}[keyof TObj & string]
