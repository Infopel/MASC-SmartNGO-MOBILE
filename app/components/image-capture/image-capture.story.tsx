import * as React from "react"
import { storiesOf } from "@storybook/react-native"
import { StoryScreen, Story, UseCase } from "../../../storybook/views"
import { color } from "../../theme"
import { ImageCapture } from "./image-capture.component"

storiesOf("ImageCapture", module)
  .addDecorator((fn) => <StoryScreen>{fn()}</StoryScreen>)
  .add("Style Presets", () => (
    <Story>
      <UseCase text="Primary" usage="The primary.">
        <ImageCapture style={{ backgroundColor: color.error }} />
      </UseCase>
    </Story>
  ))
