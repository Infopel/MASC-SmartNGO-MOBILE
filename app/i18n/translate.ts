import {TranslateOptions} from 'i18n-js/typings';
import {i18n} from './i18n';
import {TxKeyPath} from './i18n';

/**
 * Translates text.
 *
 * @param key The i18n key.
 */
export function translate(key: TxKeyPath, options?: TranslateOptions) {
  if (!key) throw new Error('Key cannot be nil or empty');
  const trans = i18n.t(key, options);

  if (typeof trans !== 'string')
  throw new Error("The inserted translation key should not map to and object");
  
  return trans;
}
