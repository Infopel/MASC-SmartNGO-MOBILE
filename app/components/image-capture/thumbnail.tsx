import { Icon } from "components/icon/icon";
import { translate } from "i18n/translate";
import { isNil } from "lodash";
import { AspectRatio, IconButton, Image, Pressable, usePropsResolution, ZStack } from "native-base";
import React from "react";

const sizes = {
    sm: '16',
    lg: '48',
}

export interface ThumbnailProps {
    src: string;
    size: keyof typeof sizes
    selected?: boolean;
    onPress?: () => void;
    onRemovePressed?: () => void;
}
export function Thumbnail({ src, selected = false, onPress, onRemovePressed, size = 'sm' }: ThumbnailProps) {
    const _size = sizes[size]

    const props = onPress ? { onPress, accessibilityRole: 'imagebutton' } : {};
    const imageAlt = translate(onPress ? "common.thumbnail" : 'common.preview-image')
    const removeIconImageAlt = (translate('common.remove'))

    const { ...resolvedProps } = usePropsResolution('button', props, { isDisabled: isNil(onPress) })
    return (<Pressable  {...resolvedProps} >
        <ZStack size={_size} >
            <AspectRatio>
                <Image size={_size} alt={translate(imageAlt)} bg='yellow.300' source={{ uri: src }} accessibilityRole='image' accessibilityState={{ selected }} />
            </AspectRatio>
            {onRemovePressed && <IconButton accessibilityLabel={removeIconImageAlt} onPress={onRemovePressed} icon={<Icon name='trash-can' type='material-community' />} />}

        </ZStack>
    </Pressable>)
}