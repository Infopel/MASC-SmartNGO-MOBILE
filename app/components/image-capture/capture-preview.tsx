import { Center, FlatList, Image, Modal, View } from 'native-base'
import React from 'react'
import { EmptyThumbnail } from './empty-thumbnail'
import { Thumbnail } from './thumbnail'
interface CapturePreviewProps {
    images: string[]
    title: string
    isVisible: boolean
    onChangeImages: (images: string[]) => void
    onClose: () => void
}
export function CapturePreview({ isVisible, images, onChangeImages, onClose, title }: CapturePreviewProps) {
    const [selectedIndex, setSelectedIndex] = React.useState(0)
    const handleOnRemove = (index: number) => {
        const items = [...images]
        items.splice(index, 1)
        onChangeImages(items)
    }


    const selectedImage = images[selectedIndex]

    return (<Modal accessibilityState={{ expanded: isVisible }} _backdrop={{}} isOpen={isVisible}>
        <Modal.Content>

            <Modal.CloseButton testID='close-previewer' onPress={onClose} />
            <Modal.Header>{title}</Modal.Header>
            <Modal.Body>
                <Center size='64' mb='1' borderRadius={8} borderColor='primary.100' borderWidth={1}>
                    {selectedImage ?
                        <Image defaultSource={require('assets/images/logo.png')} accessibilityRole='image' source={{ uri: selectedImage }} size='64' /> : <EmptyThumbnail size='lg' />}
                </Center>
                <FlatList horizontal data={images} ListEmptyComponent={() => <EmptyThumbnail size='sm'/>} ItemSeparatorComponent={() => <View w='1' />} renderItem={(({ item, index }) => <Thumbnail size='sm' src={item} selected={selectedIndex === index} onRemovePressed={handleOnRemove.bind(this, index)} onPress={setSelectedIndex.bind(this, index)} />)} />
            </Modal.Body>
        </Modal.Content>
    </Modal >)
}