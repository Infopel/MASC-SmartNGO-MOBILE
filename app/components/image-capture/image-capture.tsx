import { Box, FlatList, HStack, IBoxProps, useDisclose, VStack } from "native-base"
import * as React from "react"
import { openCamera } from 'react-native-image-crop-picker'
import logger from "utils/logger"
import { CaptureButton } from "./capture-button"
import { CapturePreview } from "./capture-preview"
import { Thumbnail } from "./thumbnail"

export interface ImageCaptureProps extends IBoxProps {
  /**
   * An optional style override useful for padding & margin.
   */
  images: string[]
  title:string
  onChangeImages: (images: string[]) => void
}

/**
 * Describe your component here
 */
export function ImageCapture(props: ImageCaptureProps) {
  const { style, images, onChangeImages,title, ...rest } = props
  const handleOnPressAdd = () => {
    openCamera({
      multiple: true,
      mediaType: 'photo',
    }).then((res) => {
      if (Array.isArray(res)) {

        props.onChangeImages([...images, ...res.map((image) => image.path)])
      } else {
        props.onChangeImages([...images, res.path])

      }
    }).catch((error) => {
      logger.log(error)
    })
  }
  const { isOpen, onOpen, onClose } = useDisclose()

  return (
    <Box {...rest}>
      <VStack p='2' borderWidth='1' borderRadius={6}  borderColor='primary.300'>
        <FlatList horizontal flex={'1'} mb='1' data={images} renderItem={(({ item }) => <Thumbnail src={item} size='sm' />)} />
        <CaptureButton onPressAdd={handleOnPressAdd} onPressPreview={onOpen} />
      </VStack>
      {isOpen && <CapturePreview title={title} onClose={onClose} isVisible={isOpen} images={images} onChangeImages={props.onChangeImages} />}
    </Box>
  )
}
