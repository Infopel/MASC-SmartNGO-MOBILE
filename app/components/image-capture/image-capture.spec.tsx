import { translate } from "i18n/translate"
import * as React from "react"
import { } from 'react-native-image-crop-picker'
import { fireEvent, render, waitFor, within } from "utils/test-utils/test-utils"
import { ImageCapture } from "./image-capture"

describe("<ImageCapture>", () => {
  const images = ['simple-image', 'simple-image2', 'simple-image3', 'simple-image4']
  const title = "Simple image capture"
  const onChangeMock = jest.fn();
  it("render", async function () {
    const { toJSON } = render(<ImageCapture title={title} onChangeImages={onChangeMock} images={[]}></ImageCapture>)
    expect(toJSON()).toMatchSnapshot()
  })

  it("should render no image", () => {
    const { getByText } = render(<ImageCapture title={title} onChangeImages={onChangeMock} images={[]}></ImageCapture>)
    expect(getByText((translate('common.camera-add')))).toBeTruthy()
    expect(getByText((translate('common.camera-preview')))).toBeTruthy()


  })

  it("should render image thumbnail with 1 image", () => {
    const _images = [images[0]]
    const { getAllByRole, debug } = render(<ImageCapture title={title} onChangeImages={onChangeMock} images={_images}></ImageCapture>)
    const imagesView = getAllByRole('image')

    expect(imagesView.length).toBe(_images.length)

    for (let i = 0; i < imagesView.length; i++) {
      const image = imagesView[i]
      expect(image).toHaveProp('source', { uri: _images[i] })
    }

  })
  it("should render image thumbnail with 4 image", () => {
    const { getAllByRole, debug } = render(<ImageCapture title={title} onChangeImages={onChangeMock} images={images}></ImageCapture>)
    const imagesView = getAllByRole('image')

    expect(imagesView.length).toBe(images.length)

    for (let i = 0; i < imagesView.length; i++) {
      const image = imagesView[i]
      expect(image).toHaveProp('source', { uri: images[i] })
    }

  })

  it("should render image preview with 4 image", () => {

    const { getByText, getByAccessibilityState, } = render(<ImageCapture title={title} onChangeImages={onChangeMock} images={images}></ImageCapture>)

    fireEvent.press(getByText((translate('common.camera-preview'))))
    const { getAllByRole, getByA11yState } = within(getByAccessibilityState({ expanded: true }))

    const imagesView = getAllByRole('imagebutton')
    expect(imagesView.length).toBe(images.length)
    expect(getByA11yState({ selected: true })).toHaveProp('source', { uri: images[0] })

    for (let i = 0; i < imagesView.length; i++) {
      const image = within(imagesView[i])
      expect(image.getByRole('image')).toHaveProp('source', { uri: images[i] })
    }



  })
  it("should set second image selected at preview", () => {
    const selectedIndex = 1
    const { getByText, getByAccessibilityState, } = render(<ImageCapture title={title} onChangeImages={onChangeMock} images={images}></ImageCapture>)

    fireEvent.press(getByText((translate('common.camera-preview'))))
    const { getAllByRole, getByA11yState } = within(getByAccessibilityState({ expanded: true }))

    const imagesView = getAllByRole('imagebutton')
    const path = images[selectedIndex]
    fireEvent.press(imagesView[selectedIndex])
    expect(getByA11yState({ selected: true })).toHaveProp('source', { uri: path })
  })
  it("should render add an image", async () => {
    const newImages = [...images, 'simple-image5']
    const { getByText } = render(<ImageCapture title={title} images={images} onChangeImages={onChangeMock}></ImageCapture>)

    fireEvent.press(getByText((translate('common.camera-add'))))

    await waitFor(
      () => expect(onChangeMock).toHaveBeenCalledWith(newImages)
    )

  })
  it("should render remove an image", () => {
    const removedIndex = 2

    const newImages = [...images]
    newImages.splice(removedIndex, 1)
    const { getByText, getByAccessibilityState } = render(<ImageCapture title={title} onChangeImages={onChangeMock} images={images}></ImageCapture>)

    fireEvent.press(getByText((translate('common.camera-preview'))))
    const { queryAllByRole } = within(getByAccessibilityState({ expanded: true }))

    const {getByRole} = within(queryAllByRole('imagebutton')[removedIndex]) 

    fireEvent.press(getByRole('button', {name: translate('common.remove')}))

    expect(onChangeMock).toHaveBeenCalledWith(newImages)


  })
  it("should close previewer", () => {
    const { getByText, queryByAccessibilityState, getByTestId } = render(<ImageCapture title={title} onChangeImages={onChangeMock} images={images}></ImageCapture>)

    fireEvent.press(getByText((translate('common.camera-preview'))))
    expect(queryByAccessibilityState({ expanded: true })).toBeTruthy()
    fireEvent.press(getByTestId(('close-previewer')))
    expect(queryByAccessibilityState({ expanded: true })).toBeFalsy()

  })

})
