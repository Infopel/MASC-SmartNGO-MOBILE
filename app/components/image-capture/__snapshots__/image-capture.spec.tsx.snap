// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<ImageCapture> render 1`] = `
<View
  dataSet={Object {}}
  style={Object {}}
>
  <View
    dataSet={Object {}}
    style={
      Object {
        "borderColor": "#f8a648",
        "borderRadius": 6,
        "borderWidth": 1,
        "paddingBottom": 8,
        "paddingLeft": 8,
        "paddingRight": 8,
        "paddingTop": 8,
      }
    }
  >
    <RCTScrollView
      contentContainerStyle={
        Array [
          Object {},
          Object {
            "dataSet": Object {},
          },
          Object {},
        ]
      }
      data={Array []}
      dataSet={Object {}}
      getItem={[Function]}
      getItemCount={[Function]}
      horizontal={true}
      keyExtractor={[Function]}
      onContentSizeChange={[Function]}
      onLayout={[Function]}
      onMomentumScrollBegin={[Function]}
      onMomentumScrollEnd={[Function]}
      onScroll={[Function]}
      onScrollBeginDrag={[Function]}
      onScrollEndDrag={[Function]}
      removeClippedSubviews={false}
      renderItem={[Function]}
      scrollEventThrottle={50}
      stickyHeaderIndices={Array []}
      style={
        Object {
          "flex": 1,
          "marginBottom": 4,
        }
      }
      viewabilityConfigCallbackPairs={Array []}
    >
      <View />
    </RCTScrollView>
    <View
      dataSet={Object {}}
      style={Object {}}
    />
    <View
      dataSet={Object {}}
      style={
        Object {
          "flex": 1,
          "flexDirection": "row",
        }
      }
    >
      <View
        accessibilityRole="button"
        accessible={true}
        collapsable={false}
        colorScheme="primary"
        dataSet={Object {}}
        focusable={true}
        onBlur={[Function]}
        onClick={[Function]}
        onFocus={[Function]}
        onResponderGrant={[Function]}
        onResponderMove={[Function]}
        onResponderRelease={[Function]}
        onResponderTerminate={[Function]}
        onResponderTerminationRequest={[Function]}
        onStartShouldSetResponder={[Function]}
        style={
          Object {
            "alignItems": "center",
            "backgroundColor": "#d37b16",
            "borderBottomLeftRadius": 4,
            "borderRadius": 0,
            "borderTopLeftRadius": 4,
            "flex": 1,
            "flexDirection": "row",
            "justifyContent": "center",
            "paddingBottom": 10,
            "paddingLeft": 12,
            "paddingRight": 12,
            "paddingTop": 10,
          }
        }
      >
        <View
          dataSet={Object {}}
          style={
            Object {
              "alignItems": "center",
              "flexDirection": "row",
            }
          }
          test={true}
        >
          <View
            dataSet={Object {}}
            style={Object {}}
          >
            <Text
              dataSet={Object {}}
              style={
                Object {
                  "backgroundColor": undefined,
                  "color": "#fafafa",
                  "fontFamily": undefined,
                  "fontSize": 14,
                  "fontStyle": "normal",
                  "fontWeight": "400",
                  "letterSpacing": 0,
                  "lineHeight": 21,
                  "textDecorationLine": undefined,
                }
              }
            >
              Add
            </Text>
          </View>
        </View>
      </View>
      <View
        dataSet={Object {}}
        style={
          Object {
            "width": 0,
          }
        }
      />
      <View
        accessibilityRole="button"
        accessible={true}
        collapsable={false}
        colorScheme="primary"
        dataSet={Object {}}
        focusable={true}
        onBlur={[Function]}
        onClick={[Function]}
        onFocus={[Function]}
        onResponderGrant={[Function]}
        onResponderMove={[Function]}
        onResponderRelease={[Function]}
        onResponderTerminate={[Function]}
        onResponderTerminationRequest={[Function]}
        onStartShouldSetResponder={[Function]}
        style={
          Object {
            "alignItems": "center",
            "backgroundColor": "#ffcc91",
            "borderBottomRightRadius": 4,
            "borderLeftWidth": 0,
            "borderRadius": 0,
            "borderTopRightRadius": 4,
            "flex": 1,
            "flexDirection": "row",
            "justifyContent": "center",
            "paddingBottom": 10,
            "paddingLeft": 12,
            "paddingRight": 12,
            "paddingTop": 10,
          }
        }
      >
        <View
          dataSet={Object {}}
          style={
            Object {
              "alignItems": "center",
              "flexDirection": "row",
            }
          }
          test={true}
        >
          <View
            dataSet={Object {}}
            style={Object {}}
          >
            <Text
              dataSet={Object {}}
              style={
                Object {
                  "backgroundColor": undefined,
                  "color": "#85551e",
                  "fontFamily": undefined,
                  "fontSize": 14,
                  "fontStyle": "normal",
                  "fontWeight": "400",
                  "letterSpacing": 0,
                  "lineHeight": 21,
                  "textDecorationLine": undefined,
                }
              }
            >
              Preview
            </Text>
          </View>
        </View>
      </View>
    </View>
  </View>
</View>
`;
